torch/csrc/jit/passes/create_functional_graphs.cpp

```diff
@@ -0,0 +1,233 @@
+    std::vector<Value*> graph_outputs;
+    for (auto it = reverse_iter.begin(); it != reverse_iter.end();) {
+      Node* n = *it++;
+
+      // constants get copied into the graph
+      if (n->kind() == prim::Constant || n == functional_subgraph_node) {
+        continue;
+      }
+
+      // if `n` is functional, all of its blocks will be merged into the
+      // new functional subgraph, so we only need to recurse if it is not
+      // functional
+      if (!functional_nodes_.count(n)) {
+        for (Block* b : n->blocks()) {
+          auto block_changed = CreateFunctionalGraphsImpl(b);
+          changed = block_changed && changed;
+        }
+        continue;
+      }
+
+      if (n->kind() == prim::FunctionalGraph &&
+          isEmptyFunctionalGraph(functional_subgraph_node)) {
+        functional_subgraph_node->destroy();
+        functional_subgraph_node = n;
+        continue;
+      }
+
+      changed = true;
+      if (aliasDb_->moveBeforeTopologicallyValid(n, functional_subgraph_node)) {
+        SubgraphUtils::mergeNodeIntoSubgraph(n, functional_subgraph_node);
+      } else {
+        functional_graph_nodes.emplace_back(functional_subgraph_node);
+        functional_subgraph_node =
+            graph_->createWithSubgraph(prim::FunctionalGraph)->insertAfter(n);
+        SubgraphUtils::mergeNodeIntoSubgraph(n, functional_subgraph_node);
+      }
+    }
+    functional_graph_nodes.emplace_back(functional_subgraph_node);
+
+    for (Node* functional_node : functional_graph_nodes) {
+      if (!inlineIfTooSmall(functional_node)) {
+        ConstantPooling(functional_node->g(attr::Subgraph));
+      }
+    }
+    return changed;
+  }
+
+  bool AnalyzeFunctionalSubset(Node* n) {
+    bool functional_node = true;
+
+    // Functional Graphs are not responsible for maintaining aliasing
+    // relationships. If an output of a functional graph escapes scope
+    // or is mutated then we might change semantics of the program if
+    // aliasing relationships are changed.
+    // For now, we don't allow any values which are mutated into the functional
+    // graph, and we don't allow any nodes which have outputs that escape scope.
+    // Possible Future Improvements:
+    // - allow inputs to have mutations so long as there are no mutations in the
+    // graph
+    // - allow functional graphs to have at most one value that can escape scope
+    // - allow outputs which alias the wildcard set but do not "re-escape"
+
+    for (Value* v : n->outputs()) {
+      bool has_writers = aliasDb_->hasWriters(v);
+      bool escapes_scope = aliasDb_->escapesScope(v);
+      if (has_writers) {
+        mutated_values_.insert(v);
+      }
+      functional_node = functional_node && !escapes_scope && !has_writers;
+    }
+
+    for (Block* block : n->blocks()) {
+      auto functional_block = AnalyzeFunctionalSubset(block);
+      functional_node = functional_node && functional_block;
+    }
+
+    auto inputs = n->inputs();
```

Comment:
It wasn't immediately obvious to me that `mutated_values_` will be populated with the outputs of previous nodes which are the inputs to this one. Worth a comment maybe.